### PR TITLE
Data flow: Prevent bad join-order in `pathStep`

### PR DESCRIPTION
**Pull request by hvitved**
Needed for https://github.com/github/codeql/pull/5629.

https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/993/
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1898/
https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/1300/ (exluding `google/guava` which failed to build)
https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/501/

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -2132,12 +2132,9 @@ private module Stage4 {
   }
 
   bindingset[node, cc, config]
-  private LocalCc getLocalCc(Node node, Cc cc, Configuration config) {
-    exists(Cc cc0 |
-      cc = pragma[only_bind_into](cc0) and
-      localFlowEntry(node, config) and
-      result = getLocalCallContext(cc0, getNodeEnclosingCallable(node))
-    )
+  LocalCc getLocalCc(Node node, Cc cc, Configuration config) {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll:2135]: Shouldn't this remain `private`?

Comment (hvitved) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll:2135]: Yes, that was accidentally removed during debugging. I'll re-add.

**Comment by jbj** [comment]
C++: no performance changes, no concerns.

File: docs/ql-libraries/dataflow/dataflow.md [hunk 2]
```diff
@@ -98,10 +98,6 @@ Recommendations:
   also work, but the upside of `use-use` steps is that sources defined in terms
   of variable reads just work out of the box. It also makes certain
   barrier-implementations simpler.
-* A predicate `DataFlowCallable Node::getEnclosingCallable()` is required, and in
```
Comment (aschackmull) [inline docs/ql-libraries/dataflow/dataflow.md:101]: With this deletion I think the document no longer mentions that `DataFlowCallable Node::getEnclosingCallable()` needs to exist. The doc really ought to say something about how a "scope" aka "enclosing callable" is required for all nodes.

**Review by aschackmull — APPROVED** [review]
QL change LGTM. The doc issue can be handled in a follow-up PR.

**Pull request merged**